src/Transport/Administration/SubscriptionManager.cs

```diff
@@ -14,83 +15,99 @@ class SubscriptionManager : ISubscriptionManager
             this.transportSettings = transportSettings;
-            this.administrativeClient = administrativeClient;
-            this.namespacePermissions = namespacePermissions;
+            this.client = client;
 
             subscriptionName = transportSettings.SubscriptionNamingConvention(subscribingQueue);
         }
 
         public async Task SubscribeAll(MessageMetadata[] eventTypes, ContextBag context, CancellationToken cancellationToken = default)
         {
-            await namespacePermissions.CanManage(cancellationToken).ConfigureAwait(false);
+            if (eventTypes.Length == 0)
+            {
+                return;
+            }
 
-            foreach (var eventType in eventTypes)
+            var ruleManager = client.CreateRuleManager(transportSettings.Topology.TopicToSubscribeOn, subscriptionName);
+            await using (ruleManager.ConfigureAwait(false))
             {
-                await SubscribeEvent(administrativeClient, eventType.MessageType, cancellationToken).ConfigureAwait(false);
+                if (eventTypes.Length == 1)
+                {
+                    await SubscribeEvent(ruleManager, eventTypes[0].MessageType, cancellationToken)
+                        .ConfigureAwait(false);
+                }
+                else
+                {
+                    var subscribeTasks = new List<Task>(eventTypes.Length);
+                    foreach (var eventType in eventTypes)
+                    {
+                        subscribeTasks.Add(SubscribeEvent(ruleManager, eventType.MessageType, cancellationToken));
+                    }
+                    await Task.WhenAll(subscribeTasks)
+                        .ConfigureAwait(false);
+                }
             }
         }
 
-        async Task SubscribeEvent(ServiceBusAdministrationClient client, Type eventType, CancellationToken cancellationToken)
+        async Task SubscribeEvent(ServiceBusRuleManager ruleManager, Type eventType,
+            CancellationToken cancellationToken)
         {
             var ruleName = transportSettings.SubscriptionRuleNamingConvention(eventType);
             var sqlExpression = $"[{Headers.EnclosedMessageTypes}] LIKE '%{eventType.FullName}%'";
-            var rule = new CreateRuleOptions(ruleName, new SqlRuleFilter(sqlExpression));
+
+            // on an entity with forwarding enabled it is not possible to do GetRules so we first have to delete and then create
+            // to preserve the update or create behavior
 
             try
             {
-                var existingRule = await client.GetRuleAsync(transportSettings.Topology.TopicToSubscribeOn, subscriptionName, rule.Name, cancellationToken).ConfigureAwait(false);
-
-                if (existingRule.Value.Filter.ToString() != rule.Filter.ToString())
-                {
-                    existingRule.Value.Action = rule.Action;
+                await ruleManager.DeleteRuleAsync(ruleName, cancellationToken)
```

Comment:
I get why we have to delete first based on the comment above but this feels like a potential for message loss since:

1. Now there will be a short window of time during each enpoint restart where the subscription is not in place
2. If something fails when we try to re-create the rule on 83 that failure would mean that we could potentially leave the subscription missing until the thing causing the failure is resolved

Am I reading this the wrong way?

Reply:
Yeah I wanted to discuss this with you in more details